packages/aws-cdk-lib/aws-elasticloadbalancingv2/lib/nlb/network-load-balancer.ts

```diff
@@ -196,16 +201,19 @@ export class NetworkLoadBalancer extends BaseLoadBalancer implements INetworkLoa
   public readonly metrics: INetworkLoadBalancerMetrics;
   public readonly securityGroups?: string[];
   public readonly ipAddressType?: IpAddressType;
+  public readonly connections: ec2.Connections;
 
   constructor(scope: Construct, id: string, props: NetworkLoadBalancerProps) {
     super(scope, id, props, {
       type: 'network',
-      securityGroups: props.securityGroups?.map(sg => sg.securityGroupId),
+      securityGroups: Lazy.list({
+        produce: () => this.connections.securityGroups.length >= 1 ? this.connections.securityGroups.map(sg => sg.securityGroupId) : undefined,
```

Comment:
Nice catch!! They are not same. I tested the following changes and I know it's can't change each other undefined and `[]`.

| before | after | error message |
| - | - | - |
| undefined | [] | A security group must be specified |
| undefined | [sg] | You cannot set security groups on a Network Load Balancer which was created without any security groups. |
| [] | undefined | 1 validation error detected: Value null at 'securityGroups' failed to satisfy constraint: Member must not be null |
| [] | [sg] | You cannot set security groups on a Network Load Balancer which was created without any security groups. |
| [sg] | undefined | 1 validation error detected: Value null at 'securityGroups' failed to satisfy constraint: Member must not be null |
| [sg] | [] | A security group must be specified |

Reply:
If implement feature flag, then I propose to change default security groups value.
If customer does not provide  security groups then AWS CDK creates a security group under the feature flag. If customer would not like to create security group, then customer should provide empty array.

This is similar codes, if under the feature flag.
```ts
new NetworkLoadBalancer(this, "NLB", {
  vpc,
});

declare const sg: ec2.SecurityGroup;
new NetworkLoadBalancer(this, "NLB", {
  vpc,
  securityGroups: [sg],
});
```

When customer is not under the feature flags, then securityGroups are as it is.
This is pseudocode.
```ts
produce: () => this.connections.securityGroups.length > 0 
  ? this.connections.securityGroups.map(sg => sg.securityGroupId) 
  : props.securityGroups?.map(sg => sg.securityGroupId),
```

I'll be implementing this, so please review again.

Reply:
Or, given that we need to keep the `securityGroups` property up-to-date, we could simply use that:
```ts
securityGroups: : Lazy.list({
  produce: () => this.securityGroups?.map(sg => sg.securityGroupId),
}),
```
This won't require the feature flag.

Reply:
Exactly. However, to keeping the each securityGroups is prone to bugs and difficult for operation. I would like to change securityGroups property to getter that reference `connections`.

```ts
public get securityGroups() {
  /**
   * With the implementation of IConnectable, connections security groups are now used instead of the security group passed in props,
   * however, since connections always return arrays that are not undefined and NLB does not allow mutual changes of undefined and empty array,
   * when the connections security groups is an empty array, the original security group is specified for backwards compatible. (original can be undefined or empty array)
   * https://github.com/aws/aws-cdk/pull/28494
   */
  return this.connections.securityGroups.length > 0
    ? this.connections.securityGroups.map(sg => sg.securityGroupId)
    : this.originalSecurityGroups?.map(sg => sg.securityGroupId);
}
/**
 * The security groups that were passed in Props.
 */
private readonly originalSecurityGroups?: ec2.ISecurityGroup[];

constructor(scope: Construct, id: string, props: NetworkLoadBalancerProps) {
  super(scope, id, props, {
    type: 'network',
    securityGroups: Lazy.list({ produce: () => this.securityGroups }),
    ipAddressType: props.ipAddressType,
  });
```